s390x: add virtio-blk-ccw support

jschintag opened the pull request:
The changes of this PR are

- create abstraction of bridge type
- move bridge var in qemu
- add ccw as bridge type
- add devno
- add virtio-blk-ccw support

Fixes: #1153

Signed-off-by: Alice Frosi <afrosi@de.ibm.com>
Reviewed-by: Jan Schintag <jan.schintag@de.ibm.com>

jschintag commented:
@grahamwhaley Can i get the RequestForComments Label? I can't set labels in this repo.

grahamwhaley commented:
@jschintag - sure - you could also put [RFC] in the subject title of the PR - the CI should also pick that up.
I'll also go add you to this repo/team (if I can work out how ;-) )

jschintag commented:
/test

devimc commented on virtcontainers/types/bridges.go:
```diff
@@ -20,6 +20,12 @@ const (
 	PCIE Type = "pcie"
 )
 
+const CCWBridgeMaxCapacity = 0xffff
```
sweet! a lot of devices

devimc commented on virtcontainers/types/bridges.go:
```diff
@@ -41,9 +47,11 @@ func NewBridge(bt Type, id string, address map[uint32]string) Bridge {
 	switch bt {
 	case PCI:
-		maxCapacity = pciBridgeMaxCapacity
+		maxCapacity = PCIBridgeMaxCapacity
```
`fallthrough` ?

jschintag replied:
yeah, i should change it to `fallthrough`

devimc commented on virtcontainers/qemu_arch_base.go:
```diff
@@ -461,70 +475,80 @@ func networkModelToQemuType(model NetInterworkingModel) govmmQemu.NetDeviceType
+func (q *qemuArchBase) appendNetwork(devices []govmmQemu.Device, endpoint Endpoint) []govmmQemu.Device {
+	d := genericNetwork(endpoint, q.vhost, q.nestedRun, q.networkIndex)
+	if d.ID == "" {
+		return devices
```
is this expected to happen?, if not, should we log something?

devimc commented on virtcontainers/qemu_arch_base.go:
```diff
@@ -461,70 +475,80 @@ func networkModelToQemuType(model NetInterworkingModel) govmmQemu.NetDeviceType
+func (q *qemuArchBase) appendBlockDevice(devices []govmmQemu.Device, drive config.BlockDrive) []govmmQemu.Device {
+	d := genericBlockDevice(drive, q.nestedRun)
+	if d == (govmmQemu.BlockDevice{}) {
+		return devices
```
same here

devimc commented on virtcontainers/qemu_arch_base.go:
```diff
@@ -461,70 +475,80 @@ func networkModelToQemuType(model NetInterworkingModel) govmmQemu.NetDeviceType
 }
 
-func (q *qemuArchBase) appendNetwork(devices []govmmQemu.Device, endpoint Endpoint) []govmmQemu.Device {
+func genericNetwork(endpoint Endpoint, vhost, nestedRun bool, index int) govmmQemu.NetDevice {
```
I wonder if wouldn't be better to return a pointer and in the caller check `if d == nil {`  or return an `error` like `genericImage`, what do you think ?

jschintag replied:
I think that depends if we want to `log` an `error` or not. I would prefer returning an `error` and `logging it`

jschintag replied:
I changed `genericNetwork` and `genericBlockDevice` to return an `error` on failure.

devimc commented on virtcontainers/qemu_arch_base.go:
```diff
@@ -385,28 +397,30 @@ func (q *qemuArchBase) appendBridges(devices []govmmQemu.Device) []govmmQemu.Dev
-		return devices
-	}
-
+func generic9PVolume(volume types.Volume, nestedRun bool) govmmQemu.FSDevice {
```
same here, return a pointer or `error` ?

devimc commented on virtcontainers/qemu_s390x.go:
```diff
@@ -95,18 +96,28 @@ func (q *qemuS390x) bridges(number uint32) {
+	id := "serial0"
+	addr, b, err := q.addDeviceToBridge(id, types.CCW)
+	if err != nil {
+		logrus.Errorf("Failed to append console: %v", err)
```
I think you can use the `virtLog` like in `qemu_amd64.go`, also please use `WithError(err)`

jschintag replied:
I will change it accordingly in the file

devimc commented on virtcontainers/qemu_s390x.go:
```diff
@@ -95,18 +96,28 @@ func (q *qemuS390x) bridges(number uint32) {
+	var devno string
+	devno, err = b.AddressFormatCCW(addr)
+	if err != nil {
+		logrus.Errorf("Failed to append console: %v", err)
```
`virtLog` and `WithError`

devimc commented on virtcontainers/qemu_s390x.go:
```diff
@@ -124,6 +135,32 @@ func (q *qemuS390x) appendConsole(devices []govmmQemu.Device, path string) []gov
+func (q *qemuS390x) appendImage(devices []govmmQemu.Device, path string) ([]govmmQemu.Device, error) {
+	drive, err := genericImage(path)
+	if err != nil {
+		logrus.Errorf("Failed to append image: %v", err)
```
same here

devimc commented on virtcontainers/qemu_s390x.go:
```diff
@@ -124,6 +135,32 @@ func (q *qemuS390x) appendConsole(devices []govmmQemu.Device, path string) []gov
+	d := genericBlockDevice(drive, false)
+	addr, b, err := q.addDeviceToBridge(drive.ID, types.CCW)
+	if err != nil {
+		logrus.Errorf("Failed to append blk-dev: %v", err)
```
and here and in all the `logrus.Errorf` in this file

jschintag commented on virtcontainers/kata_agent.go:
```diff
@@ -1127,22 +1131,27 @@ func (k *kataAgent) buildContainerRootfs(sandbox *Sandbox, c *Container, rootPat
 			}
 
-		} else {
+		case sandbox.config.HypervisorConfig.BlockDeviceDriver == config.VirtioSCSI:
```
It was the default before, but we could also add an error as default if no Driver is configured.

jschintag commented on virtcontainers/kata_agent.go:
```diff
@@ -1401,13 +1410,17 @@ func (k *kataAgent) handleBlockVolumes(c *Container) []*grpc.Storage {
 			vol.Driver = kataMmioBlkDevType
 			vol.Source = blockDrive.VirtPath
-		} else {
+		case c.sandbox.config.HypervisorConfig.BlockDeviceDriver == config.VirtioSCSI:
```
There should not be any unknown Drivers here

jodh-intel replied:
But adding a `default` would make the test exhaustive and flag any possible issues in the future.

In fact, looking more closely, I think `handleBlockVolumes()` should return an `error` too as it's already (trying to) handle two error scenarios which it looks like the caller ignores (since it's not obvious this function can "fail".

jschintag replied:
Added `error` to the return of `handleBlockVolumes()` and made the default for both a return with `error`.

jschintag commented on virtcontainers/types/bridges.go:
```diff
@@ -7,41 +7,59 @@ package types
+	Type Type
+
+	// BridgeMaxCapacity is the max capacity of the bridge
+	BridgeMaxCapacity uint32
```
I agree

jschintag commented on virtcontainers/qemu_arch_base.go:
```diff
@@ -96,6 +97,18 @@ type qemuArch interface {
+	// getBridges grants access to Bridges
+	getBridges() []types.Bridge
+
+	// addBridge adds a new Brdige to Bridges
```
agreed, i will change it

jschintag commented on virtcontainers/qemu_arch_base.go:
```diff
@@ -588,3 +598,49 @@ func (q *qemuArchBase) setIgnoreSharedMemoryMigrationCaps(ctx context.Context, q
+	for _, b := range q.Bridges {
+		err = b.RemoveDevice(ID)
+		if err == nil {
+			// device was removed correctly
```
There could be more than one bridge and we don't know to which bridge the device belongs to. Therefore we only return an error if we do not find it in any bridge. The error returned by `b.RemveDevice(ID)` is always `fmt.Errorf("Unable to hot unplug device %s: not present on bridge", ID)`

jschintag commented on virtcontainers/types/bridges.go:
```diff
@@ -7,41 +7,59 @@ package types
 
-	// Addr is the PCI/e slot of the bridge
+	// Addr is the slot of the bridge
 	Addr int
```
I could rename `Address` to `Devices` or `DevAddr`. Any suggestions?

jodh-intel replied:
That would work I guess. I agree with @devimc that having two fields which suggest the same "thing" is confusing and therefore error-prone.

devimc replied:
@jschintag `Devices` sounds good to me

jschintag replied:
@devimc @jodh-intel I renamed it to `Devices`

jschintag commented on virtcontainers/qemu.go:
```diff
@@ -1991,12 +1987,9 @@ func (q *qemu) load(s persistapi.HypervisorState) {
-			Addr:    bridge.Addr,
-		})
+		b := types.NewBridge(types.Type(bridge.Type), bridge.ID, bridge.DeviceAddr)
+		b.Addr = bridge.Addr
```
Yeah i add it as a parameter.

jschintag commented:
@devimc Thank you for your review

jschintag commented:
Here is what i changed:
- Changed `BridgeMaxCapacity` to `MaxCapacity`
- Added Addr as parameter to `NewBridge` and added `fallthrough` to the `switch`
- Changed description of `addBridge`
- Use `virtlog` in `qemu_s390x.go`

I also noticed that the qemu-tests for other archs weren`t included in the changes. Fixed it.

/test

devimc commented:
CI is failing, @jschintag  any thoughts?

```
command failed error 'exit status 125'
[docker run --cidfile /tmp/cid900304392/OlACPKKDhZGKDUpOEYW8wA6f2zYc74 --runtime kata-runtime --device /dev/loop0 --device /dev/loop1 --device /dev/loop2 --device /dev/loop3 --device /dev/loop4 --device /dev/loop5 --device /dev/loop6 --device /dev/loop7 --device /dev/loop8 --device /dev/loop9 --rm --name OlACPKKDhZGKDUpOEYW8wA6f2zYc74 busybox stat /dev/loop0 /dev/loop1 /dev/loop2 /dev/loop3 /dev/loop4 /dev/loop5 /dev/loop6 /dev/loop7 /dev/loop8 /dev/loop9]
Timeout: 120 seconds
Exit Code: 125
Stdout: 
Stderr: docker: Error response from daemon: OCI runtime create failed: rpc error: code = InvalidArgument desc = invalid ID and VM path for device type:"blk" container_path:"/dev/loop0": unknown.
```

jschintag commented:
@devimc I cannot replicate the issue on s390x, i need to further look into it.

jschintag commented:
@devimc As far as i can see, the agent receives a grpc Request to add a device with an empty `id`. However i can't replicate the issue on my local x86. Let's check if it reoccurs.

/test

nitkon commented:
/test-power

jschintag commented:
@grahamwhaley I can't replicate the issue on my local x86 machine.
<details>
<code><pre>
make FOCUS="run hot plug block devices" docker
ln -sf . vendor/src
GOPATH=/root/go/src/github.com/kata-containers/tests/vendor go build ./vendor/github.com/onsi/ginkgo/ginkgo
unlink vendor/src
./ginkgo -failFast -v -focus "run hot plug block devices" -skip "" \
        ./integration/docker/ -- -runtime=kata-runtime -timeout=120
Running command '/usr/local/bin/kata-runtime [kata-runtime --kata-show-default-config-paths]'
[kata-runtime --kata-show-default-config-paths]
Timeout: 120 seconds
Exit Code: 0
Stdout: /etc/kata-containers/configuration.toml
/usr/share/defaults/kata-containers/configuration.toml

Stderr:
Running Suite: Integration Suite
================================
Random Seed: 1566203206
Will run 1 of 249 specs

Running command '/usr/bin/docker [docker pull busybox]'
[docker pull busybox]
Timeout: 600 seconds
Exit Code: 0
Stdout: Using default tag: latest
latest: Pulling from library/busybox
Digest: sha256:9f1003c480699be56815db0f8146ad2e22efea85129b5b5983d0e0fb52d9ab70
Status: Image is up to date for busybox:latest
docker.io/library/busybox:latest

Stderr:
Running command '/usr/bin/docker [docker pull alpine:3.7]'
[docker pull alpine:3.7]
Timeout: 600 seconds
Exit Code: 0
Stdout: 3.7: Pulling from library/alpine
5d20c808ce19: Pulling fs layer
5d20c808ce19: Verifying Checksum
5d20c808ce19: Download complete
5d20c808ce19: Pull complete
Digest: sha256:8421d9a84432575381bfabd248f1eb56f3aa21d9d7cd2511583c68c9b7511d10
Status: Downloaded newer image for alpine:3.7
docker.io/library/alpine:3.7

Stderr:
Running command '/usr/bin/docker [docker pull debian]'
[docker pull debian]
Timeout: 600 seconds
Exit Code: 0
Stdout: Using default tag: latest
latest: Pulling from library/debian
4ae16bd47783: Pulling fs layer
4ae16bd47783: Verifying Checksum
4ae16bd47783: Download complete
4ae16bd47783: Pull complete
Digest: sha256:2f04d3d33b6027bb74ecc81397abe780649ec89f1a2af18d7022737d0482cefe
Status: Downloaded newer image for debian:latest
docker.io/library/debian:latest

Stderr:
Running command '/usr/bin/docker [docker pull fedora]'
[docker pull fedora]
Timeout: 600 seconds
Exit Code: 0
Stdout: Using default tag: latest
latest: Pulling from library/fedora
fd2e8b5b2254: Pulling fs layer
fd2e8b5b2254: Download complete
fd2e8b5b2254: Pull complete
Digest: sha256:d39a02a0f13c1df3bbcb0ccea4021c53b8e0bfd87f701a5115e18ec089814e70
Status: Downloaded newer image for fedora:latest
docker.io/library/fedora:latest

Stderr:
Running command '/usr/bin/docker [docker pull centos]'
[docker pull centos]
Timeout: 600 seconds
Exit Code: 0
Stdout: Using default tag: latest
latest: Pulling from library/centos
8ba884070f61: Pulling fs layer
8ba884070f61: Verifying Checksum
8ba884070f61: Download complete
8ba884070f61: Pull complete
Digest: sha256:a799dd8a2ded4a83484bbae769d97655392b3f86533ceb7dd96bbac929809f3c
Status: Downloaded newer image for centos:latest
docker.io/library/centos:latest

Stderr:
Running command '/usr/bin/docker [docker pull vish/stress]'
[docker pull vish/stress]
Timeout: 600 seconds
Exit Code: 0
Stdout: Using default tag: latest
latest: Pulling from vish/stress
2ce2382a5646: Pulling fs layer
2ce2382a5646: Verifying Checksum
2ce2382a5646: Download complete
2ce2382a5646: Pull complete
Digest: sha256:b6456a3df6db5e063e1783153627947484a3db387be99e49708c70a9a15e7177
Status: Downloaded newer image for vish/stress:latest
docker.io/vish/stress:latest

Stderr:
Running command '/usr/bin/docker [docker images --format {{.Repository}}]'
[docker images --format {{.Repository}}]
Timeout: 120 seconds
Exit Code: 0
Stdout: debian
fedora
busybox
alpine
helloworld
edgemicro
node
centos
alpine
vish/stress

Stderr:
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
run hot plug block devices
  should be attached
  /root/go/src/github.com/kata-containers/tests/integration/docker/run_test.go:147
Running command '/bin/dd [dd if=/dev/zero of=/tmp/dd899282237 count=1 bs=50M]'
[dd if=/dev/zero of=/tmp/dd899282237 count=1 bs=50M]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr: 1+0 records in
1+0 records out
52428800 bytes (52 MB, 50 MiB) copied, 0.101386 s, 517 MB/s

Running command '/bin/bash [bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd899282237]'
[bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd899282237]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Welcome to fdisk (util-linux 2.31.1).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Device does not contain a recognized partition table.
Created a new DOS disklabel with disk identifier 0xcf30aab8.

Command (m for help): Created a new GPT disklabel (GUID: 43C93B83-4C5A-E443-932C-6CE4BC8E7C64).

Command (m for help): Partition number (1-128, default 1): First sector (2048-102366, default 2048): Last sector, +sectors or +size{K,M,G,T,P} (2048-102366, default 102366):
Created a new partition 1 of type 'Linux filesystem' and of size 49 MiB.

Command (m for help): The partition table has been altered.
Syncing disks.


Stderr:
Running command '/sbin/losetup [losetup -fP /tmp/dd899282237]'
[losetup -fP /tmp/dd899282237]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -j /tmp/dd899282237]'
[losetup -j /tmp/dd899282237]
Timeout: 120 seconds
Exit Code: 0
Stdout: /dev/loop2: [64514]:131098 (/tmp/dd899282237)

Stderr:
Running command '/bin/dd [dd if=/dev/zero of=/tmp/dd906912632 count=1 bs=50M]'
[dd if=/dev/zero of=/tmp/dd906912632 count=1 bs=50M]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr: 1+0 records in
1+0 records out
52428800 bytes (52 MB, 50 MiB) copied, 0.108625 s, 483 MB/s

Running command '/bin/bash [bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd906912632]'
[bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd906912632]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Welcome to fdisk (util-linux 2.31.1).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Device does not contain a recognized partition table.
Created a new DOS disklabel with disk identifier 0x2b86c9d4.

Command (m for help): Created a new GPT disklabel (GUID: 08352D12-43FB-344E-8171-5547A0912355).

Command (m for help): Partition number (1-128, default 1): First sector (2048-102366, default 2048): Last sector, +sectors or +size{K,M,G,T,P} (2048-102366, default 102366):
Created a new partition 1 of type 'Linux filesystem' and of size 49 MiB.

Command (m for help): The partition table has been altered.
Syncing disks.


Stderr:
Running command '/sbin/losetup [losetup -fP /tmp/dd906912632]'
[losetup -fP /tmp/dd906912632]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -j /tmp/dd906912632]'
[losetup -j /tmp/dd906912632]
Timeout: 120 seconds
Exit Code: 0
Stdout: /dev/loop3: [64514]:131101 (/tmp/dd906912632)

Stderr:
Running command '/bin/dd [dd if=/dev/zero of=/tmp/dd211389047 count=1 bs=50M]'
[dd if=/dev/zero of=/tmp/dd211389047 count=1 bs=50M]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr: 1+0 records in
1+0 records out
52428800 bytes (52 MB, 50 MiB) copied, 0.0871715 s, 601 MB/s

Running command '/bin/bash [bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd211389047]'
[bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd211389047]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Welcome to fdisk (util-linux 2.31.1).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Device does not contain a recognized partition table.
Created a new DOS disklabel with disk identifier 0x346a642a.

Command (m for help): Created a new GPT disklabel (GUID: DD638E6C-9EDF-3349-B0F5-A627ECE9B46A).

Command (m for help): Partition number (1-128, default 1): First sector (2048-102366, default 2048): Last sector, +sectors or +size{K,M,G,T,P} (2048-102366, default 102366):
Created a new partition 1 of type 'Linux filesystem' and of size 49 MiB.

Command (m for help): The partition table has been altered.
Syncing disks.


Stderr:
Running command '/sbin/losetup [losetup -fP /tmp/dd211389047]'
[losetup -fP /tmp/dd211389047]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -j /tmp/dd211389047]'
[losetup -j /tmp/dd211389047]
Timeout: 120 seconds
Exit Code: 0
Stdout: /dev/loop4: [64514]:131102 (/tmp/dd211389047)

Stderr:
Running command '/bin/dd [dd if=/dev/zero of=/tmp/dd006137706 count=1 bs=50M]'
[dd if=/dev/zero of=/tmp/dd006137706 count=1 bs=50M]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr: 1+0 records in
1+0 records out
52428800 bytes (52 MB, 50 MiB) copied, 0.115736 s, 453 MB/s

Running command '/bin/bash [bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd006137706]'
[bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd006137706]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Welcome to fdisk (util-linux 2.31.1).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Device does not contain a recognized partition table.
Created a new DOS disklabel with disk identifier 0x962282a9.

Command (m for help): Created a new GPT disklabel (GUID: 82FB0021-6BFE-4F48-A0A1-EE7D562E9C63).

Command (m for help): Partition number (1-128, default 1): First sector (2048-102366, default 2048): Last sector, +sectors or +size{K,M,G,T,P} (2048-102366, default 102366):
Created a new partition 1 of type 'Linux filesystem' and of size 49 MiB.

Command (m for help): The partition table has been altered.
Syncing disks.


Stderr:
Running command '/sbin/losetup [losetup -fP /tmp/dd006137706]'
[losetup -fP /tmp/dd006137706]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -j /tmp/dd006137706]'
[losetup -j /tmp/dd006137706]
Timeout: 120 seconds
Exit Code: 0
Stdout: /dev/loop5: [64514]:131103 (/tmp/dd006137706)

Stderr:
Running command '/bin/dd [dd if=/dev/zero of=/tmp/dd946753985 count=1 bs=50M]'
[dd if=/dev/zero of=/tmp/dd946753985 count=1 bs=50M]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr: 1+0 records in
1+0 records out
52428800 bytes (52 MB, 50 MiB) copied, 0.0763301 s, 687 MB/s

Running command '/bin/bash [bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd946753985]'
[bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd946753985]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Welcome to fdisk (util-linux 2.31.1).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Device does not contain a recognized partition table.
Created a new DOS disklabel with disk identifier 0xc2e293e8.

Command (m for help): Created a new GPT disklabel (GUID: 118FD8D8-36AC-844E-80BE-2C94F0F78A00).

Command (m for help): Partition number (1-128, default 1): First sector (2048-102366, default 2048): Last sector, +sectors or +size{K,M,G,T,P} (2048-102366, default 102366):
Created a new partition 1 of type 'Linux filesystem' and of size 49 MiB.

Command (m for help): The partition table has been altered.
Syncing disks.


Stderr:
Running command '/sbin/losetup [losetup -fP /tmp/dd946753985]'
[losetup -fP /tmp/dd946753985]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -j /tmp/dd946753985]'
[losetup -j /tmp/dd946753985]
Timeout: 120 seconds
Exit Code: 0
Stdout: /dev/loop6: [64514]:131104 (/tmp/dd946753985)

Stderr:
Running command '/bin/dd [dd if=/dev/zero of=/tmp/dd383974444 count=1 bs=50M]'
[dd if=/dev/zero of=/tmp/dd383974444 count=1 bs=50M]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr: 1+0 records in
1+0 records out
52428800 bytes (52 MB, 50 MiB) copied, 0.10151 s, 516 MB/s

Running command '/bin/bash [bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd383974444]'
[bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd383974444]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Welcome to fdisk (util-linux 2.31.1).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Device does not contain a recognized partition table.
Created a new DOS disklabel with disk identifier 0x0dcee642.

Command (m for help): Created a new GPT disklabel (GUID: AA7BF5B9-57AB-9B4A-86D0-2E74AF8967E9).

Command (m for help): Partition number (1-128, default 1): First sector (2048-102366, default 2048): Last sector, +sectors or +size{K,M,G,T,P} (2048-102366, default 102366):
Created a new partition 1 of type 'Linux filesystem' and of size 49 MiB.

Command (m for help): The partition table has been altered.
Syncing disks.


Stderr:
Running command '/sbin/losetup [losetup -fP /tmp/dd383974444]'
[losetup -fP /tmp/dd383974444]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -j /tmp/dd383974444]'
[losetup -j /tmp/dd383974444]
Timeout: 120 seconds
Exit Code: 0
Stdout: /dev/loop7: [64514]:131105 (/tmp/dd383974444)

Stderr:
Running command '/bin/dd [dd if=/dev/zero of=/tmp/dd826486171 count=1 bs=50M]'
[dd if=/dev/zero of=/tmp/dd826486171 count=1 bs=50M]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr: 1+0 records in
1+0 records out
52428800 bytes (52 MB, 50 MiB) copied, 0.0748785 s, 700 MB/s

Running command '/bin/bash [bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd826486171]'
[bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd826486171]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Welcome to fdisk (util-linux 2.31.1).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Device does not contain a recognized partition table.
Created a new DOS disklabel with disk identifier 0x151ff382.

Command (m for help): Created a new GPT disklabel (GUID: 91160215-78DD-7E41-A203-C9F384F3EBDC).

Command (m for help): Partition number (1-128, default 1): First sector (2048-102366, default 2048): Last sector, +sectors or +size{K,M,G,T,P} (2048-102366, default 102366):
Created a new partition 1 of type 'Linux filesystem' and of size 49 MiB.

Command (m for help): The partition table has been altered.
Syncing disks.


Stderr:
Running command '/sbin/losetup [losetup -fP /tmp/dd826486171]'
[losetup -fP /tmp/dd826486171]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -j /tmp/dd826486171]'
[losetup -j /tmp/dd826486171]
Timeout: 120 seconds
Exit Code: 0
Stdout: /dev/loop12: [64514]:131106 (/tmp/dd826486171)

Stderr:
Running command '/bin/dd [dd if=/dev/zero of=/tmp/dd026542654 count=1 bs=50M]'
[dd if=/dev/zero of=/tmp/dd026542654 count=1 bs=50M]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr: 1+0 records in
1+0 records out
52428800 bytes (52 MB, 50 MiB) copied, 0.0720971 s, 727 MB/s

Running command '/bin/bash [bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd026542654]'
[bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd026542654]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Welcome to fdisk (util-linux 2.31.1).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Device does not contain a recognized partition table.
Created a new DOS disklabel with disk identifier 0x0703b3df.

Command (m for help): Created a new GPT disklabel (GUID: C09F796B-38AB-C64A-9F99-3D67159D3D66).

Command (m for help): Partition number (1-128, default 1): First sector (2048-102366, default 2048): Last sector, +sectors or +size{K,M,G,T,P} (2048-102366, default 102366):
Created a new partition 1 of type 'Linux filesystem' and of size 49 MiB.

Command (m for help): The partition table has been altered.
Syncing disks.


Stderr:
Running command '/sbin/losetup [losetup -fP /tmp/dd026542654]'
[losetup -fP /tmp/dd026542654]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -j /tmp/dd026542654]'
[losetup -j /tmp/dd026542654]
Timeout: 120 seconds
Exit Code: 0
Stdout: /dev/loop13: [64514]:131107 (/tmp/dd026542654)

Stderr:
Running command '/bin/dd [dd if=/dev/zero of=/tmp/dd430980229 count=1 bs=50M]'
[dd if=/dev/zero of=/tmp/dd430980229 count=1 bs=50M]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr: 1+0 records in
1+0 records out
52428800 bytes (52 MB, 50 MiB) copied, 0.104901 s, 500 MB/s

Running command '/bin/bash [bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd430980229]'
[bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd430980229]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Welcome to fdisk (util-linux 2.31.1).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Device does not contain a recognized partition table.
Created a new DOS disklabel with disk identifier 0xbdfaccfe.

Command (m for help): Created a new GPT disklabel (GUID: 2DC26E3E-40C7-844D-AF45-29494742F171).

Command (m for help): Partition number (1-128, default 1): First sector (2048-102366, default 2048): Last sector, +sectors or +size{K,M,G,T,P} (2048-102366, default 102366):
Created a new partition 1 of type 'Linux filesystem' and of size 49 MiB.

Command (m for help): The partition table has been altered.
Syncing disks.


Stderr:
Running command '/sbin/losetup [losetup -fP /tmp/dd430980229]'
[losetup -fP /tmp/dd430980229]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -j /tmp/dd430980229]'
[losetup -j /tmp/dd430980229]
Timeout: 120 seconds
Exit Code: 0
Stdout: /dev/loop14: [64514]:131108 (/tmp/dd430980229)

Stderr:
Running command '/bin/dd [dd if=/dev/zero of=/tmp/dd416564768 count=1 bs=50M]'
[dd if=/dev/zero of=/tmp/dd416564768 count=1 bs=50M]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr: 1+0 records in
1+0 records out
52428800 bytes (52 MB, 50 MiB) copied, 0.119541 s, 439 MB/s

Running command '/bin/bash [bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd416564768]'
[bash -c printf "g\nn\n\n\n\nw\n" | fdisk /tmp/dd416564768]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Welcome to fdisk (util-linux 2.31.1).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Device does not contain a recognized partition table.
Created a new DOS disklabel with disk identifier 0x736b2f41.

Command (m for help): Created a new GPT disklabel (GUID: 9FFF66B6-5A16-1747-ACC3-8ABE06EE1646).

Command (m for help): Partition number (1-128, default 1): First sector (2048-102366, default 2048): Last sector, +sectors or +size{K,M,G,T,P} (2048-102366, default 102366):
Created a new partition 1 of type 'Linux filesystem' and of size 49 MiB.

Command (m for help): The partition table has been altered.
Syncing disks.


Stderr:
Running command '/sbin/losetup [losetup -fP /tmp/dd416564768]'
[losetup -fP /tmp/dd416564768]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -j /tmp/dd416564768]'
[losetup -j /tmp/dd416564768]
Timeout: 120 seconds
Exit Code: 0
Stdout: /dev/loop15: [64514]:131109 (/tmp/dd416564768)

Stderr:
Running command '/usr/bin/docker [docker run --cidfile /tmp/cid082798550/0cbAcnASystCAH5314i6Jw3Bgk2ER1 --runtime kata-runtime --device /dev/loop2 --device /dev/loop3 --device /dev/loop4 --device /dev/loop5 --device /dev/loop6 --device /dev/loop7 --device /dev/loop12 --device /dev/loop13 --device /dev/loop14 --device /dev/loop15 --rm --name 0cbAcnASystCAH5314i6Jw3Bgk2ER1 busybox stat /dev/loop2 /dev/loop3 /dev/loop4 /dev/loop5 /dev/loop6 /dev/loop7 /dev/loop12 /dev/loop13 /dev/loop14 /dev/loop15]'
[docker run --cidfile /tmp/cid082798550/0cbAcnASystCAH5314i6Jw3Bgk2ER1 --runtime kata-runtime --device /dev/loop2 --device /dev/loop3 --device /dev/loop4 --device /dev/loop5 --device /dev/loop6 --device /dev/loop7 --device /dev/loop12 --device /dev/loop13 --device /dev/loop14 --device /dev/loop15 --rm --name 0cbAcnASystCAH5314i6Jw3Bgk2ER1 busybox stat /dev/loop2 /dev/loop3 /dev/loop4 /dev/loop5 /dev/loop6 /dev/loop7 /dev/loop12 /dev/loop13 /dev/loop14 /dev/loop15]
Timeout: 120 seconds
Exit Code: 0
Stdout:   File: /dev/loop2
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 23h/35d Inode: 1144        Links: 1     Device type: fe,1
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2019-08-19 08:27:35.000000000
Modify: 2019-08-19 08:27:35.000000000
Change: 2019-08-19 08:27:35.000000000

  File: /dev/loop3
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 23h/35d Inode: 1145        Links: 1     Device type: fe,11
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2019-08-19 08:27:35.000000000
Modify: 2019-08-19 08:27:35.000000000
Change: 2019-08-19 08:27:35.000000000

  File: /dev/loop4
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 23h/35d Inode: 1146        Links: 1     Device type: fe,21
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2019-08-19 08:27:35.000000000
Modify: 2019-08-19 08:27:35.000000000
Change: 2019-08-19 08:27:35.000000000

  File: /dev/loop5
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 23h/35d Inode: 1147        Links: 1     Device type: fe,30
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2019-08-19 08:27:35.000000000
Modify: 2019-08-19 08:27:35.000000000
Change: 2019-08-19 08:27:35.000000000

  File: /dev/loop6
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 23h/35d Inode: 1148        Links: 1     Device type: fe,40
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2019-08-19 08:27:35.000000000
Modify: 2019-08-19 08:27:35.000000000
Change: 2019-08-19 08:27:35.000000000

  File: /dev/loop7
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 23h/35d Inode: 1149        Links: 1     Device type: fe,51
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2019-08-19 08:27:35.000000000
Modify: 2019-08-19 08:27:35.000000000
Change: 2019-08-19 08:27:35.000000000

  File: /dev/loop12
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 23h/35d Inode: 1150        Links: 1     Device type: fe,61
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2019-08-19 08:27:35.000000000
Modify: 2019-08-19 08:27:35.000000000
Change: 2019-08-19 08:27:35.000000000

  File: /dev/loop13
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 23h/35d Inode: 1151        Links: 1     Device type: fe,70
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2019-08-19 08:27:35.000000000
Modify: 2019-08-19 08:27:35.000000000
Change: 2019-08-19 08:27:35.000000000

  File: /dev/loop14
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 23h/35d Inode: 1152        Links: 1     Device type: fe,80
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2019-08-19 08:27:35.000000000
Modify: 2019-08-19 08:27:35.000000000
Change: 2019-08-19 08:27:35.000000000

  File: /dev/loop15
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 23h/35d Inode: 1153        Links: 1     Device type: fe,90
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2019-08-19 08:27:35.000000000
Modify: 2019-08-19 08:27:35.000000000
Change: 2019-08-19 08:27:35.000000000


Stderr:
Running command '/usr/bin/docker [docker ps -a -f name=0cbAcnASystCAH5314i6Jw3Bgk2ER1 --format {{.Status}}]'
[docker ps -a -f name=0cbAcnASystCAH5314i6Jw3Bgk2ER1 --format {{.Status}}]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -d /dev/loop2]'
[losetup -d /dev/loop2]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -d /dev/loop3]'
[losetup -d /dev/loop3]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -d /dev/loop4]'
[losetup -d /dev/loop4]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -d /dev/loop5]'
[losetup -d /dev/loop5]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -d /dev/loop6]'
[losetup -d /dev/loop6]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -d /dev/loop7]'
[losetup -d /dev/loop7]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -d /dev/loop12]'
[losetup -d /dev/loop12]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -d /dev/loop13]'
[losetup -d /dev/loop13]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -d /dev/loop14]'
[losetup -d /dev/loop14]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:
Running command '/sbin/losetup [losetup -d /dev/loop15]'
[losetup -d /dev/loop15]
Timeout: 120 seconds
Exit Code: 0
Stdout:
Stderr:

• [SLOW TEST:9.975 seconds]
run
/root/go/src/github.com/kata-containers/tests/integration/docker/run_test.go:101
  hot plug block devices
  /root/go/src/github.com/kata-containers/tests/integration/docker/run_test.go:146
    should be attached
    /root/go/src/github.com/kata-containers/tests/integration/docker/run_test.go:147
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSS
Ran 1 of 249 Specs in 44.077 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 248 Skipped
PASS

Ginkgo ran 1 suite in 50.775969517s
Test Suite Passed
</pre>
</details>

grahamwhaley commented:
/test
@jschintag - let me re-trigger the CIs and see what happens. There were quite a few failing though.
/cc @devimc

jschintag commented:
/test

I rebased the PR to master to solve the Merge conflicts

jodh-intel commented on virtcontainers/types/bridges.go:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	if addr == 0 {
+		return 0, fmt.Errorf("Unable to hot plug device on bridge: there are not empty slots")
```
Nit: "there are no empty slots".

jschintag replied:
Fixed it, thanks.

jodh-intel commented on virtcontainers/types/bridges.go:
```diff
@@ -0,0 +1,114 @@
+// AddressFormatCCW returns the address format for the device number. The channel subsystem-ID 0xfe is reserved to the virtual channel and the address format is in the form fe.n.dddd, where n is subchannel set ID and ddd the device number. More details at https://www.ibm.com/support/knowledgecenter/en/linuxonibm/com.ibm.linux.z.ldva/ldva_t_configuringSCSIdevices.html
+func (b *Bridge) AddressFormatCCW(addr string) (string, error) {
+	if b.Type != CCW {
+		return "", fmt.Errorf("Wrong bridge type")
```
Maybe,

```go
return "", fmt.Errorf("Expected bridge type %T, got %T (%+v)", CCW, b.Type, b)
```

jschintag replied:
I changed it to your suggestion, thanks.

jodh-intel commented on virtcontainers/types/bridges.go:
```diff
@@ -0,0 +1,114 @@
+	if b.Type != CCW {
+		return "", fmt.Errorf("Wrong bridge type")
+	}
+	return fmt.Sprintf("fe.%x.%s", b.Addr, addr), nil
```
Nit: Missing blank line.

jschintag replied:
Fixed it, thanks.

devimc commented:
CI still failing :worried: 

```
Stderr: docker: Error response from daemon: OCI runtime create failed: rpc error: code = InvalidArgument desc = invalid ID and VM path for device type:"blk" container_path:"/dev/loop0": unknown.
```

devimc commented:
/test-ubuntu

jschintag commented:
/test

jschintag commented:
/test-ubuntu

Let's check if this fixes the netmon error

jschintag commented:
/test

jschintag commented:
/test

some of the failures seem to be from the CI

jschintag commented:
/test-ubuntu

jschintag commented:
/test

jschintag commented:
/test

jschintag commented:
/test

devimc commented:
@jschintag I got these logs from the proxy's journal

```
pci 0000:01:09.0: [1af4:1001] type 00 class 0x010000
pci 0000:01:09.0: reg 0x10: [io  0x0000-0x007f]
pci 0000:01:09.0: reg 0x14: [mem 0x00000000-0x00000fff]
pci 0000:01:09.0: reg 0x20: [mem 0x00000000-0x00003fff 64bit pref]
pci 0000:01:09.0: BAR 4: assigned [mem 0xfe820000-0xfe823fff 64bit pref]
pci 0000:01:09.0: BAR 1: assigned [mem 0xfe408000-0xfe408fff]
pci 0000:01:09.0: BAR 0: no space for [io  size 0x0080]
pci 0000:01:09.0: BAR 0: failed to assign [io  size 0x0080]
virtio-pci 0000:01:09.0: enabling device (0000 -> 0002)
```

seems like the kernel is unable to online it

devimc commented on virtcontainers/qemu.go:
```diff
@@ -1050,26 +1020,43 @@ func (q *qemu) hotplugAddBlockDevice(drive *config.BlockDrive, op operation, dev
 		}()
 
-		// PCI address is in the format bridge-addr/device-addr eg. "03/02"
-		drive.PCIAddr = fmt.Sprintf("%02x", bridge.Addr) + "/" + addr
```
I think that you're not honouring this format, maybe for this the CI is failing ? :thinking:

jschintag replied:
That could be, Let's try it out.

jschintag replied:
Thanks, that was the issue.

jschintag commented:
/test

devimc commented:
@jschintag initrd CI is happy, why did you remove that line?

jschintag commented:
@devimc I actually don't know why that line was removed. Alice wrote that part, i'll ask her on Monday. I'll also look into the opensuse-15 failure.

devimc commented:
@jschintag suse has been failing, it's not stable right now

jschintag commented:
@devimc so, do you think we can merge this PR?

alicefr commented:
> @devimc I actually don't know why that line was removed. Alice wrote that part, i'll ask her on Monday. I'll also look into the opensuse-15 failure.

My mistake. I did something wrong during the rebase probably sorry. Thx @jschintag

jschintag commented:
/test-ubuntu

jschintag commented:
I just rebased the PR to the current master.

devimc commented:
adding dnm label, I'd like to review this again

alicefr commented:
@devimc please let us know if we need to split this PR is smallers PR. Without these changes the kata runtime is broken on s390x.

devimc commented:
/test

jschintag commented:
@jodh-intel What are your thoughts on this. Is this PR fine as it is, or do we need to split it into smaller PRs?

jodh-intel commented:
Ok - silence is compliance... merging.
